src/validation.cpp

```diff
@@ -4955,19 +4955,35 @@ void ChainstateManager::CheckBlockIndex()
         // Chainstate-specific checks on setBlockIndexCandidates
         for (auto c : GetAll()) {
             if (c->m_chain.Tip() == nullptr) continue;
+            // Two main factors determine whether pindex is a candidate in
+            // setBlockIndexCandidates:
+            //
+            // - If pindex has less work than the chain tip, it should not be a
+            //   candidate, and this will be asserted below. Otherwise it is a
+            //   potential candidate.
+            //
+            // - If pindex or one of its parent blocks never downloaded
+            //   transactions (pindexFirstNeverProcessed is non-null), it should
+            //   not be a candidate, and this will be asserted below. Otherwise
+            //   it is a potential candidate.
             if (!CBlockIndexWorkComparator()(pindex, c->m_chain.Tip()) && pindexFirstNeverProcessed == nullptr) {
+                // If pindex was detected as invalid (pindexFirstInvalid is
+                // non-null), it is not required to be in
+                // setBlockIndexCandidates.
                 if (pindexFirstInvalid == nullptr) {
-                    const bool is_active = c == &ActiveChainstate();
-                    // If this block sorts at least as good as the current tip and
-                    // is valid and we have all data for its parents, it must be in
-                    // setBlockIndexCandidates.  m_chain.Tip() must also be there
-                    // even if some data has been pruned.
+                    // If pindex and all its parents downloaded transactions, and
+                    // the transactions were not pruned (pindexFirstMissing is
+                    // null), it is a potential candidate.
```

Comment:
49426f84b9a15aaae1b3a3c9c5d2ac4ee5855b11: the description matches the code, but I'm a bit confused why an index with a pruned ancestor can't be a candidate. Or is it the case that it _may_ be a candidate, but we just don't have a check for that?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520231273

> [49426f8](https://github.com/bitcoin/bitcoin/commit/49426f84b9a15aaae1b3a3c9c5d2ac4ee5855b11): the description matches the code, but I'm a bit confused why an index with a pruned ancestor can't be a candidate. Or is it the case that it _may_ be a candidate, but we just don't have a check for that?

I think it's the latter. It may be a candidate but we can't have a check ensuring that it is a candidate because if blocks are missing between pindex and the current chain tip, FindMostWorkChain could remove it from the candidate set:

https://github.com/bitcoin/bitcoin/blob/a945f09fa6e0f94cc424da9e06516f9cfa192545/src/validation.cpp#L3073

Probably ideally the check below would be `pindexFirstMissingSinceChainTip == nullptr` instead of `pindexFirstMissing == nullptr`, but it would take more work to be able to write that.

Reply:
That makes sense. So instead of checking if there is a block with missing data between the tip and the new candidate, the code here checks if there's any missing data between genesis and the new candidate. Which is typically the case on a pruned node, which means this check is useless on a pruned node. But that's outside the scope of this PR.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520231273

> That makes sense. So instead of checking if there is a block with missing data between the tip and the new candidate, the code here checks if there's any missing data between genesis and the new candidate.

Yes, it's pretty confusing so I added some more comments to help clarify.